File: pulumi/examples/go-foundation/1-org/essential_contacts.go
/*
 * Copyright 2026 Vitruvian Software
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package main

import (
	"fmt"

	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/essentialcontacts"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// deployEssentialContacts creates organization-level Essential Contacts
// for notification routing. This mirrors the Terraform foundation's
// essential_contacts.tf, implementing the full category-to-group mapping:
//
//	BILLING         → billing_admins, billing_data_users
//	LEGAL           → org_admins, audit_data_users
//	PRODUCT_UPDATES → org_admins
//	SECURITY        → scc_admin, security_reviewer
//	SUSPENSION      → org_admins
//	TECHNICAL       → security_reviewer, network_viewer
//
// Uses the configurable language tag (H9, defaults to "en").
func deployEssentialContacts(ctx *pulumi.Context, cfg *OrgConfig) error {
	parent := cfg.Parent // "organizations/<id>" or "folders/<id>"
	lang := cfg.EssentialContactsLanguage

	// Build a map of email → notification categories, mirroring the TF
	// transpose(categories_map) pattern. This groups categories per email
	// so each contact gets a single resource with all their categories.
	contactMap := make(map[string][]string)

	addContact := func(email string, categories ...string) {
		if email == "" {
			return
		}
		contactMap[email] = append(contactMap[email], categories...)
	}

	// Map categories to groups exactly as TF does.
	// TF's categories_map uses fallbacks to org_admins when optional groups
	// are not configured. We replicate this behavior here.

	// Resolve group emails with fallbacks (G12)
	sccAdmin := cfg.GCPSCCAdmin
	if sccAdmin == "" {
		sccAdmin = cfg.GroupOrgAdmins // fallback
	}
	securityReviewer := cfg.GCPSecurityReviewer
	if securityReviewer == "" {
		securityReviewer = cfg.GroupOrgAdmins // fallback
	}
	networkViewer := cfg.GCPNetworkViewer
	if networkViewer == "" {
		networkViewer = cfg.GroupOrgAdmins // fallback
	}

	// BILLING → billing_admins + billing_data_users
	addContact(cfg.GroupBillingAdmins, "BILLING")
	addContact(cfg.BillingDataUsers, "BILLING")

	// LEGAL → org_admins + audit_data_users
	addContact(cfg.GroupOrgAdmins, "LEGAL")
	addContact(cfg.AuditDataUsers, "LEGAL")

	// PRODUCT_UPDATES → org_admins
	addContact(cfg.GroupOrgAdmins, "PRODUCT_UPDATES")

	// SECURITY → scc_admin (or org_admins fallback) + security_reviewer (or org_admins fallback)
	addContact(sccAdmin, "SECURITY")
	addContact(securityReviewer, "SECURITY")

	// SUSPENSION → org_admins
	addContact(cfg.GroupOrgAdmins, "SUSPENSION")

	// TECHNICAL → security_reviewer (or org_admins) + network_viewer (or org_admins)
	addContact(securityReviewer, "TECHNICAL")
	addContact(networkViewer, "TECHNICAL")

	// De-duplicate categories per contact
	for email, cats := range contactMap {
		seen := make(map[string]bool)
		unique := make([]string, 0, len(cats))
		for _, c := range cats {
			if !seen[c] {
				seen[c] = true
				unique = append(unique, c)
			}
		}
		contactMap[email] = unique
	}

	// Create one Essential Contact per unique email
	idx := 0
	for email, categories := range contactMap {
		catArray := make(pulumi.StringArray, len(categories))
		for i, c := range categories {
			catArray[i] = pulumi.String(c)
		}

		if _, err := essentialcontacts.NewContact(ctx, fmt.Sprintf("essential-contact-%d", idx), &essentialcontacts.ContactArgs{
			Parent:                            pulumi.String(parent),
			Email:                             pulumi.String(email),
			LanguageTag:                       pulumi.String(lang),
			NotificationCategorySubscriptions: catArray,
		}); err != nil {
			return err
		}
		idx++
	}

	return nil
}

File: pulumi/examples/ts-foundation/1-org/essential_contacts.ts
/**
 * Copyright 2026 Vitruvian Software
 *
 * 1-org/essential_contacts.ts — Organization Essential Contacts.
 * Mirrors: 1-org/envs/shared/essential_contacts.tf and
 * 1-org/essential_contacts.go in the Go foundation.
 *
 * Creates organization-level Essential Contacts for notification routing,
 * mapping notification categories (BILLING, LEGAL, SECURITY, etc.) to
 * the appropriate Google Groups.
 */

import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";
import { OrgConfig } from "./config";

/**
 * deployEssentialContacts creates Essential Contacts resources.
 * Port of Go foundation's deployEssentialContacts (essential_contacts.go:38-127).
 *
 * Category → Group mapping (mirrors TF's transpose(categories_map)):
 *   BILLING         → billing_admins, billing_data_users
 *   LEGAL           → org_admins, audit_data_users
 *   PRODUCT_UPDATES → org_admins
 *   SECURITY        → scc_admin (fallback: org_admins), security_reviewer (fallback: org_admins)
 *   SUSPENSION      → org_admins
 *   TECHNICAL       → security_reviewer (fallback: org_admins), network_viewer (fallback: org_admins)
 */
export function deployEssentialContacts(cfg: OrgConfig): void {
    const parent = cfg.parent; // "organizations/<id>" or "folders/<id>"
    const lang = cfg.essentialContactsLanguage;

    // Build contact map: email → notification categories
    const contactMap: Record<string, Set<string>> = {};

    const addContact = (email: string | undefined, ...categories: string[]) => {
        if (!email) return;
        if (!contactMap[email]) contactMap[email] = new Set();
        for (const cat of categories) {
            contactMap[email].add(cat);
        }
    };

    // Resolve group emails with fallbacks (matches Go G12 pattern)
    const orgAdmins = cfg.requiredGroups.group_org_admins;
    const sccAdmin = cfg.gcpGroups.sccAdmin || orgAdmins;
    const securityReviewer = cfg.gcpGroups.securityReviewer || orgAdmins;
    const networkViewer = cfg.gcpGroups.networkViewer || orgAdmins;

    // BILLING → billing_admins + billing_data_users
    addContact(cfg.requiredGroups.group_billing_admins, "BILLING");
    addContact(cfg.requiredGroups.billing_data_users, "BILLING");

    // LEGAL → org_admins + audit_data_users
    addContact(orgAdmins, "LEGAL");
    addContact(cfg.requiredGroups.audit_data_users, "LEGAL");

    // PRODUCT_UPDATES → org_admins
    addContact(orgAdmins, "PRODUCT_UPDATES");

    // SECURITY → scc_admin + security_reviewer
    addContact(sccAdmin, "SECURITY");
    addContact(securityReviewer, "SECURITY");

    // SUSPENSION → org_admins
    addContact(orgAdmins, "SUSPENSION");

    // TECHNICAL → security_reviewer + network_viewer
    addContact(securityReviewer, "TECHNICAL");
    addContact(networkViewer, "TECHNICAL");

    // Create one Essential Contact per unique email
    let idx = 0;
    for (const [email, categories] of Object.entries(contactMap)) {
        new gcp.essentialcontacts.Contact(`essential-contact-${idx}`, {
            parent,
            email,
            languageTag: lang,
            notificationCategorySubscriptions: [...categories],
        });
        idx++;
    }
}
